lib/compress/drop-side-effect-free.js

```diff
@@ -121,7 +121,7 @@ def_drop_side_effect_free(AST_Constant, return_null);
 def_drop_side_effect_free(AST_This, return_null);
 
 def_drop_side_effect_free(AST_Call, function (compressor, first_in_statement) {
-    if (this.optional && is_nullish(this.expression, compressor)) {
+    if (is_nullish(this, compressor)) {
         return make_node(AST_Undefined, this);
     }
```

Comment:
This condition works 100% as far as I can tell, but it leaves me uncomfortable.

If we change `is_nullish` in such a way that it can find out of an AST_Call, AST_Dot or AST_Sub can be nullish themselves or not, this introduces an issue: What if the AST_Call, AST_Dot, or AST_Sub were nullish because they were `null`, but we're returning `undefined`? 

I'd say, we need a way to figure out if something is actually null, or undefined. Maybe `is_nullish` could return `"null" | "undefined"` instead of `true | false`.

Reply:
And I don't mean to be snob, I mean I totally failed to understand how chains work in the first place :D

Reply:
> What if the AST_Call, AST_Dot, or AST_Sub were nullish because they were null? In that case, we're returning undefined here.

I'm not sure I understand? `undefined?.foo` and `null?.foo` both evaluate to `undefined`, `null` isn't propagated.

Are you instead saying this would incorrectly drop a intermediate chain location? Eg `foo.returns_null_but_isnt_pure()?.bar`. In this case, we could turn this into just `foo.returns_null_but_isnt_pure()`, but that would need to be done at the `AST_Chain` level instead of here.

Reply:
I mean, if is_nullish says that

`a.b.c` "is nullish"

Then that non-optional chain could return null or undefined. But we're only returning undefined.

Indeed if it's an optional chain the result is always undefined, but it might not be an optional one.

Reply:
I believe what you're saying is that, if we were to find `a?.b` is nullish, we want to leave the nullish part? Eg, if `a = null`, then we want to drop the expression down to the `null` value. Or if `a = undefined`, we want to drop it down to the `undefined` value.

I update the code to be

```js
if (is_nullish()) {
  return this.expression.drop_side_effect_free()
}
```

Which I believe captures your case.

Reply:
That's not what I mean. I mean that if `a.b.c` is nullish (because `a = { b: { c: null } }`), it doesn't mean that it returns undefined. It can be null too as in this example.

I don't think it's a problem right now though.

Reply:
Terser could, feasibly, understand that `a.b.c` is "nullish", which would break this code. But this is more of a future concern than a present one.

Reply:
Notice I'm not using `?.`

Reply:
Ok, I updated this to check explicitly for a short-circuited optional chain expression.

Reply:
💪 looking great